### PR TITLE
GUACAMOLE-474: Allow file uploads and downloads to be disabled

### PR DESCRIPTION
**Pull request by necouchman**
This change adds options for the protocols that support various file transfer mechanisms to disable either uploads or downloads (or both) if so desired.

I took the approach of creating new boolean options for this, which seems the simplest approach; however, I did think of one possible alternative approach which would be to change the current "Enable Drive" (RDP) and "Enable SFTP" (RDP, SSH, and VNC) options such that, instead of boolean (checkbox), they are multi-option (select) with enable, upload-only, download-only, or disabled.  Let me know if that seems a better approach.

### CODE REVIEW

**Comment by ibuler** [comment]
Can dispatch a event when upload or download ?

**Comment by mike-jumper** [comment]
Looks like this will need some rebasing due to #243.

**Comment by necouchman** [comment]
@mike-jumper: Rebase completed.

File: src/common-ssh/sftp.c [hunk 1]
```diff
@@ -787,6 +787,13 @@ static int guac_common_ssh_sftp_get_handler(guac_user* user,
     /* Otherwise, send file contents */
     else {
 
+        /* If downloads are disabled, log and return. */
+        if (filesystem->disable_download) {
+            guac_user_log(user, GUAC_LOG_INFO,
+                    "File downloads have been disabled.");
```
Comment (mike-jumper) [inline src/common-ssh/sftp.c:793]: In line with the other download failure messages (`Unable to read file "..."`, `Unable to read directory "...": (reason)`), I suggest phrasing this message as a failed/refused operation and including the name of the file.

The current message (`File downloads have been disabled.`) is not immediately clear that it indicates that a request isn't being serviced, and could be interpreted as a simple acknowledgement of the download setting.

File: src/protocols/rdp/channels/rdpdr/rdpdr-fs-messages.c [hunk 2]
```diff
@@ -98,9 +98,13 @@ void guac_rdpdr_fs_process_create(guac_rdp_common_svc* svc,
         /* Create \Download if it doesn't exist */
         file = guac_rdp_fs_get_file((guac_rdp_fs*) device->data, file_id);
         if (file != NULL && strcmp(file->absolute_path, "\\") == 0) {
-            int download_id =
-                guac_rdp_fs_open((guac_rdp_fs*) device->data, "\\Download",
-                    GENERIC_READ, 0, FILE_OPEN_IF, FILE_DIRECTORY_FILE);
+            
+            /* Only create Download folder if downloads are enabled. */
+            if(!((guac_rdp_fs*) devices->data)->disable_download) {
```
Comment (mike-jumper) [inline src/protocols/rdp/channels/rdpdr/rdpdr-fs-messages.c:103]: Whoops - need that space after `if`.

File: src/protocols/rdp/fs.c [hunk 3]
```diff
@@ -77,11 +79,15 @@ void guac_rdp_fs_free(guac_rdp_fs* fs) {
...
 guac_object* guac_rdp_fs_alloc_object(guac_rdp_fs* fs, guac_user* user) {
-
+    
     /* Init filesystem */
     guac_object* fs_object = guac_user_alloc_object(user);
     fs_object->get_handler = guac_rdp_download_get_handler;
-    fs_object->put_handler = guac_rdp_upload_put_handler;
+    
+    /* Assign handler only if uploads are not disabled. */
+    if(!fs->disable_upload)
```
Comment (mike-jumper) [inline src/protocols/rdp/fs.c:88]: `if(` -> `if (`

;)

File: src/protocols/rdp/channels/rdpdr/rdpdr-fs-messages.c (hunk 2 above)
Comment (necouchman) [reply src/protocols/rdp/channels/rdpdr/rdpdr-fs-messages.c:103]: Agh!  I'm slipping into old habits...

**Pull request merged**